awscli/autocomplete/indexer.py

```diff
@@ -0,0 +1,118 @@
+# Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+#     http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+import logging
+import sqlite3
+
+
+LOG = logging.getLogger(__name__)
+
+
+class ModelIndexer(object):
+    def __init__(self, db_connection):
+        self._db_connection = db_connection
+
+    def generate_index(self, clidriver):
+        parent = 'aws'
+        self._db_connection.execute(
+            'INSERT OR REPLACE INTO command_table (command)'
+            'VALUES (:command)',
+            command=parent
+        )
+        command_table = clidriver.subcommand_table
+        self._generate_arg_index(command=parent, parent='',
+                                 arg_table=clidriver.arg_table)
+        self._generate_command_index(command_table, parent=parent)
+
+    def _generate_arg_index(self, command, parent, arg_table):
+        for name, value in arg_table.items():
+            self._db_connection.execute(
+                'INSERT INTO param_table '
+                '(argname, type_name, command, parent, nargs)'
+                ' VALUES (:argname, :type_name, :command, :parent, :nargs)',
+                argname=name, type_name=value.cli_type_name,
+                command=command, parent=parent,
+                nargs=value.nargs,
+            )
+
+    def _generate_command_index(self, command_table, parent):
+        for name, command in command_table.items():
+            self._db_connection.execute(
+                'INSERT INTO command_table (command, parent) '
+                'VALUES (:command, :parent)',
+                command=name, parent=parent,
+            )
+            self._generate_arg_index(command=name, parent=parent,
+                                     arg_table=command.arg_table)
+            self._generate_command_index(command.subcommand_table,
+                                         parent='%s.%s' % (parent, name))
+
+
+# This is similar to DBConnection in awscli.customization.history.
```

Comment:
It's also going to be challenging to remember to enforce that you can't import anything "heavyweight" in a `dbutils` module.  It's much easier to keep track of the fact that anything in `awscli/autocomplete/` should not import anything like `clidriver/botocore`.

It will also depend on how much logic is similar.  They've already diverged a bit and may likely diverge further.  Right now we'd only save about 10-20 lines of code.

Reply:
>> It's also going to be challenging to remember to enforce that you can't import anything "heavyweight" in a dbutils module... Right now we'd only save about 10-20 lines of code

That's fair enough. I wonder if there are other ways we can ensure that we do not import anything heavy weight into this module. An easy win would be put a comment at the top of the file that says to minimize imports (right now you have to scroll midway through the file to realize it). But I do wonder if there is any tests/checks (I would need to brainstorm examples) we can put in place for additional ensurance our imports are lightweight?

Reply:
Yeah I definitely have a few ideas, I just think they're much lower priority than getting everything implemented first.